internal/store/store.go

```diff
@@ -487,3 +500,49 @@ func (s *Store) GetImage(composeID uuid.UUID) (*Image, error) {
 
 	return nil, &NotFoundError{"compose could not be found"}
 }
+
+func (s *Store) PushSource(source SourceConfig) {
+	s.change(func() error {
+		s.Sources[source.Name] = source
+		return nil
+	})
+}
+
+func (s *Store) DeleteSource(name string) {
+	s.change(func() error {
+		delete(s.Sources, name)
+		return nil
+	})
+}
+
+func (s *Store) ListSources() []string {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	names := make([]string, 0, len(s.Sources))
+	for name := range s.Sources {
+		names = append(names, name)
+	}
+	sort.Strings(names)
+
+	return names
+}
+
+func (s *Store) GetSource(name string, source *SourceConfig) bool {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	var ok bool
+	*source, ok = s.Sources[name]
+	if !ok {
+		return false
+	}
+
+	return true
+}
+
+func (s *Store) GetAllSources() map[string]SourceConfig {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.Sources
+}
```

Comment:
I am uncertain about what to do with this function. This will return a reference which means that the store's sources can be modified anywhere this gets called. Do you have any recommendations on how to guarantee isolation for the store's sources?

Reply:
Return copies. It may feel less efficient, but you are right, if we return a reference then the lock must be held while the reference is live, which we don't want to expose to the caller.